prometheus/src/main/java/com/uber/m3/tally/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,434 @@
+ *      Thread.sleep(500);
+ *   }
+ * }
+ * </pre>
+ */
+public class PrometheusReporter implements StatsReporter {
+
+    static final String METRIC_ID_KEY_VALUE = "1";
+    private static final String PREFIX_SPLITTER = "+";
+    private static final String KEY_PAIR_SPLITTER = ",";
+    private static final String KEY_NAME_SPLITTER = "=";
+    private static final String KEY_PAIR_TEMPLATE = "%s" + KEY_NAME_SPLITTER + "%s";
+    private static final TimerType DEFAULT_TIMER_TYPE = TimerType.SUMMARY;
+    private static final String[] EMPTY_STRING_ARRAY = new String[0];
+
+    private final CollectorRegistry registry;
+    private final TimerType timerType;
+    private final Map<Double, Double> defaultQuantiles;
+    private final double[] defaultBuckets;
+    private final int ageBuckets;
+    private final long maxAgeSeconds;
+    private final ConcurrentMap<String, Counter> registeredCounters;
+    private final ConcurrentMap<String, Gauge> registeredGauges;
+    private final ConcurrentMap<String, Histogram> registeredHistograms;
+    private final ConcurrentMap<String, Summary> registeredSummaries;
+
+    private PrometheusReporter(
+            Map<Double, Double> defaultQuantiles,
+            double[] defaultBuckets,
+            TimerType defaultTimerType,
+            CollectorRegistry registry,
+            int ageBuckets,
+            long maxAgeSeconds
+    ) {
+        this.registry = registry;
+        this.timerType = defaultTimerType;
+        this.defaultBuckets = defaultBuckets;
+        this.defaultQuantiles = defaultQuantiles;
+        this.ageBuckets = ageBuckets;
+        this.maxAgeSeconds = maxAgeSeconds;
+        this.registeredCounters = new ConcurrentHashMap<>();
+        this.registeredGauges = new ConcurrentHashMap<>();
+        this.registeredSummaries = new ConcurrentHashMap<>();
+        this.registeredHistograms = new ConcurrentHashMap<>();
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public void reportCounter(String name, Map<String, String> tags, long value) {
+        final Map<String, String> finalTags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, finalTags.keySet());
+        registeredCounters.computeIfAbsent(collectorName, key -> Counter.build()
+                .name(name)
+                .help(String.format("%s counter", name))
+                .labelNames(collectionToStringArray(finalTags.keySet()))
+                .register(registry))
+                .labels(collectionToStringArray(finalTags.values())).inc(value);
+    }
+
+    @Override
+    public void reportGauge(String name, Map<String, String> tags, double value) {
+        final Map<String, String> finalTags = (tags == null) ? Collections.emptyMap() : tags;
+        String collectorName = canonicalMetricId(name, finalTags.keySet());
+        registeredGauges.computeIfAbsent(collectorName, key -> Gauge.build()
```

Comment:
Nit: I think this might read more clearly if we separate out the `computeIfAbsent` call from the updating of the value, e.g.:

```
gauge = registeredGauges.computeIfAbsent(...)

gauge.labels(...).set(value)
```
Makes it a bit more clear what happens once and what happens every call.